spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java

```diff
@@ -81,23 +82,17 @@
 
   @Override
   CloseableIterator<InternalRow> open(FileScanTask task) {
+    SparkDeleteFilter deletes = new SparkDeleteFilter(task, tableSchema, expectedSchema);
+
+    // schema or rows returned by readers
+    Schema requiredSchema = deletes.requiredSchema();
+    Map<Integer, ?> idToConstant = PartitionUtil.constantsMap(task, RowDataReader::convertConstant);
     DataFile file = task.file();
 
     // update the current file for Spark's filename() function
     InputFileBlockHolder.set(file.path().toString(), task.start(), task.length());
 
-    // schema or rows returned by readers
-    PartitionSpec spec = task.spec();
-    Set<Integer> idColumns = spec.identitySourceIds();
-    Schema partitionSchema = TypeUtil.select(expectedSchema, idColumns);
-    boolean projectsIdentityPartitionColumns = !partitionSchema.columns().isEmpty();
-
-    if (projectsIdentityPartitionColumns) {
-      return open(task, expectedSchema, PartitionUtil.constantsMap(task, RowDataReader::convertConstant))
-          .iterator();
-    }
-    // return the base iterator
-    return open(task, expectedSchema, ImmutableMap.of()).iterator();
+    return deletes.filter(open(task, requiredSchema, idToConstant)).iterator();
```

Comment:
The same question similar to https://github.com/apache/iceberg/pull/1444/files#r489309503

Reply:
See the other thread.